### PR TITLE
[postgresql] Fix checkPrimaryKeyUnicity option

### PR DESCRIPTION
**Pull request by elpaso**
This provider option was linked to the project level option
"Trust layer metadata..." which was implemented
to speed up loading of large datasets by trusting extent
read from metadata to avoid costly operations to determine
the layer extent.

Check PK unicity on the other hand has only effect on views
and query layers and it is useful as an independent
option to prevent loading of layers that have no PK (or the
wrong one).

But the operation of determine unicity of a values in a column
can also be costly, so better to get control back to the user.

Legacy default is preserved (the project-level "Trust..." option).

Fixes #21839

**Funded by RAAP.nl**

Contextual tooltips have been added to explain the option, the option is only available when it makes sense (on views and materialized views).

Screenshot_20190426_092736

Screenshot_20190426_092754

### CODE REVIEW

**Comment by elpaso** [comment]
@pblottiere I'd like to know your opinion on this PR, I couldn't find a way to test if `v.in.ogr` is broken, there was a comment in the code about removing `checkPrimaryKeyUnicity` from the URI, if that's still an issue I think we should clean the URI downstream.

File: src/ui/qgsprojectpropertiesbase.ui [hunk 1]
```diff
@@ -1378,7 +1378,7 @@
            <item row="3" column="0">
             <widget class="QCheckBox" name="mTrustProjectCheckBox">
              <property name="toolTip">
-              <string>Speed up project loading by skipping data checks. Useful in qgis server context or project with huge database views or materialized views.</string>
+              <string>Speed up project loading by skipping data checks in Postgres layers. Useful in QGIS server context or project with huge database views or materialized views.</string>
```
Comment (jef-n) [inline src/ui/qgsprojectpropertiesbase.ui]: PostgreSQL

File: src/providers/postgres/qgspgtablemodel.cpp [hunk 2]
```diff
@@ -35,8 +36,11 @@ QgsPgTableModel::QgsPgTableModel()
   headerLabels << tr( "SRID" );
   headerLabels << tr( "Feature id" );
   headerLabels << tr( "Select at id" );
+  headerLabels << tr( "Check PK unicity" );
   headerLabels << tr( "Sql" );
   setHorizontalHeaderLabels( headerLabels );
+  setHeaderData( 8, Qt::Orientation::Horizontal, tr( "Disable 'Fast Access to Features at ID' capability to force keeping the attribute table in memory (e.g. in case of expensive views)." ), Qt::ToolTipRole );
```
Comment (jef-n) [inline src/providers/postgres/qgspgtablemodel.cpp]: QgsPgTableModel::DbtmSelectAtId

**Comment by jef-n** [comment]
Ignoring the option was IMHO a bad idea anyway - adding a layer with a plugin would use the flag, saving and reloading would suddenly ignore it.  I would have just tied it back to `mUseEstimatedMetadata` although that should meanwhile better be called `mTradeAccuracyForPerformance`.

**Comment by elpaso** [comment]
@jef-n thanks for the review!

**Comment by pblottiere** [comment]
> Ignoring the option was IMHO a bad idea anyway 

I agree, it's much better this way, thanks @elpaso!

**Comment by pblottiere** [comment]
>  I'd like to know your opinion on this PR, I couldn't find a way to test if v.in.ogr is broken, there was a comment in the code about removing checkPrimaryKeyUnicity from the URI, if that's still an issue I think we should clean the URI downstream.

This comment dates from 2015 so we could definitely hope than it has been fixed in the meantime :).

By blaming the sources, it seems that @jef-n added this comment. What do you think about it?

**Comment by jef-n** [comment]
> This comment dates from 2015 so we could definitely hope than it has been fixed in the meantime :).
> By blaming the sources, it seems that @jef-n added this comment. What do you think about it?

Nope, from 2008 in [r8654](https://trac.osgeo.org/qgis/changeset/8654/trunk) (aka 409dcee2c65f) - but yes, this should not be a issue anymore - I think this was about calling v.in.ogr from the GRASS plugin back when processing didn't even exists.

File: tests/src/python/test_provider_postgres.py [hunk 3]
```diff
@@ -1097,6 +1097,37 @@ def testReadExtentOnTable(self):
...
+        # This is NOT valid because the default is to check unicity
+        vl0 = QgsVectorLayer(self.dbconn + ' sslmode=disable key=\'an_int\' srid=0 type=POINT table="qgis_test"."b21839_pk_unicity_view" (geom) sql=', 'test', 'postgres')
+        self.assertFalse(vl0.isValid())
+
+        # This is valid because the readExtentFromXml option is set
+        options = QgsVectorLayer.LayerOptions(True, True)  # loadDefaultStyle, readExtentFromXml
+        vl0 = QgsVectorLayer(self.dbconn + ' sslmode=disable key=\'an_int\' srid=0 type=POINT table="qgis_test"."b21839_pk_unicity_view" (geom) sql=', 'test', 'postgres', options)
+        self.assertTrue(vl0.isValid())
+
+        # Valid because a_unique_int is unique
```
Comment (luipir) [inline tests/src/python/test_provider_postgres.py]:
```suggestion
        # Valid because a_unique_int is unique and default is to check unicity
```

**Pull request merged**